aiida/engine/daemon/execmanager.py

```diff
@@ -196,13 +192,16 @@ def find_data_node(inputs, uuid):
         if data_node is None:
             logger.warning('failed to load Node<{}> specified in the `local_copy_list`'.format(uuid))
         else:
-            # Note, once #2579 is implemented, use the `node.open` method instead of the named temporary file in
-            # combination with the new `Transport.put_object_from_filelike`
-            # Since the content of the node could potentially be binary, we read the raw bytes and pass them on
-            with NamedTemporaryFile(mode='wb+') as handle:
-                handle.write(data_node.get_object_content(filename, mode='rb'))
-                handle.flush()
-                transport.put(handle.name, target)
+            with folder.open(target, 'wb') as handle:
+                with data_node.open(filename, 'rb') as source:
+                    shutil.copyfileobj(source, handle)
+            provenance_exclude_list.append(target)
+
+    # In a dry_run, the working directory is the raw input folder, which will already contain these resources
+    if not dry_run:
+        for filename in folder.get_content_list():
+            logger.debug('[submission of calculation {}] copying file/folder {}...'.format(node.pk, filename))
+            transport.put(folder.get_abs_path(filename), filename)
 
     if dry_run:
         if remote_copy_list:
```

Comment:
So, you moved this here and now you have a `dry_run` check right next to another; would it perhaps be better to unify these two conditional branches? (the first is just `if not dry_run`, but the second has an `else` clause in line 225).

Reply:
Done and also added a test for the nested paths